scipy/integrate/_cubature.py

```diff
@@ -600,82 +600,74 @@ def __init__(self, f, a, b, xp):
         self._orig_a = a
         self._orig_b = b
 
-        self._semi_inf_pos = []
-        self._double_inf_pos = []
+        # (-oo, oo) will be mapped to (-1, 1).
+        self._double_inf_pos = (a == -math.inf) & (b == math.inf)
 
-        ndim = xp_size(a)
-
-        for i in range(ndim):
-            if a[i] == -math.inf and b[i] == math.inf:
-                # (-oo, oo) will be mapped to (-1, 1).
-                self._double_inf_pos.append(i)
-
-            elif a[i] != -math.inf and b[i] == math.inf:
-                # (start, oo) will be mapped to (0, 1).
-                self._semi_inf_pos.append(i)
+        # (start, oo) will be mapped to (0, 1).
+        start_inf_mask = (a != -math.inf) & (b == math.inf)
 
-            elif a[i] == -math.inf and b[i] != math.inf:
-                # (-oo, end) will be mapped to (0, 1).
-                #
-                # This is handled by making the transformation t = -x and reducing it to
-                # the other semi-infinite case.
-                self._semi_inf_pos.append(i)
+        # (-oo, end) will be mapped to (0, 1).
+        inf_end_mask = (a == -math.inf) & (b != math.inf)
+        # This is handled by making the transformation t = -x and reducing it to
+        # the other semi-infinite case.
+        self._semi_inf_pos = start_inf_mask | inf_end_mask
 
-                # Since we flip the limits, we don't need to separately multiply the
-                # integrand by -1.
-                self._orig_a[i] = -b[i]
-                self._orig_b[i] = -a[i]
+        # Since we flip the limits, we don't need to separately multiply the
+        # integrand by -1.
+        self._orig_a[inf_end_mask] = -b[inf_end_mask]
+        self._orig_b[inf_end_mask] = -a[inf_end_mask]
 
     @property
     def transformed_limits(self):
         a = xp_copy(self._orig_a)
         b = xp_copy(self._orig_b)
 
-        for index in self._double_inf_pos:
-            a[index] = -1
-            b[index] = 1
+        a[self._double_inf_pos] = -1
+        b[self._double_inf_pos] = 1
 
-        for index in self._semi_inf_pos:
-            a[index] = 0
-            b[index] = 1
+        a[self._semi_inf_pos] = 0
+        b[self._semi_inf_pos] = 1
 
         return a, b
 
     @property
     def points(self):
         # If there are infinite limits, then the origin becomes a problematic point
         # due to a division by zero there.
-        if len(self._double_inf_pos) != 0 or len(self._semi_inf_pos) != 0:
+        if (self._xp.sum(self._double_inf_pos) != 0
```

Comment:
This change makes array api strict unhappy because it sums a non-numeric dtype. Probably needs something like
```suggestion
        if (self._xp.sum(self._double_inf_pos.astype(xp.int32)) != 0
```
or similar